chain/neutrino_test.go

```diff
@@ -0,0 +1,190 @@
+// the client into the scanning state and that subsequent calls while scanning
+// will call Update on the client's Rescanner.
+func TestNeutrinoClientNotifyReceived(t *testing.T) {
+	var (
+		timeout                 = time.After(1 * time.Second)
+		nc                      = newMockNeutrinoClient()
+		wantNotifyReceivedCalls = 50
+		wantUpdateCalls         = wantNotifyReceivedCalls - 1
+	)
+
+	// executeCalls calls NotifyReceived() synchronously n times without
+	// blocking the test and requires no error after each call.
+	executeCalls := func(n int) <-chan struct{} {
+		var (
+			addrs []btcutil.Address
+			done  = make(chan struct{})
+		)
+
+		go func() {
+			defer close(done)
+
+			for i := 0; i < n; i++ {
+				err := nc.NotifyReceived(addrs)
+				require.NoError(t, err)
+			}
+		}()
+
+		return done
+	}
+
+	// Wait for all calls to complete or test to time out.
+	select {
+	case <-timeout:
+		t.Fatal("timed out")
+	case <-executeCalls(wantNotifyReceivedCalls):
+		// Require that the expected number of calls to Update were made
+		// once done sending all NotifyReceived calls.
+		mockRescan := nc.rescan.(*mockRescanner)
+		gotUpdateCalls := mockRescan.updateArgs.Len()
+		require.Equal(t, wantUpdateCalls, gotUpdateCalls)
+	}
+}
+
+// TestNeutrinoClientNotifyReceivedRescan verifies concurrent calls to
+// NotifyBlocks, NotifyReceived and Rescan do not result in a data race
+// and that there is no panic on replacing the rescan goroutine single instance.
+func TestNeutrinoClientNotifyReceivedRescan(t *testing.T) {
+	var (
+		addrs        []btcutil.Address
+		nc           = newMockNeutrinoClient()
+		timeout      = time.After(1 * time.Second)
+		wantRoutines = 100
+	)
+
+	// Define closures to wrap desired neutrino client method calls.
+
+	// callRescan calls the Rescan() method and asserts it completes
+	// with no errors. Rescan() is called with the hash of an empty header
+	// on each call.
+	var (
+		emptyHeader = &wire.BlockHeader{}
+		startHash   = emptyHeader.BlockHash()
+	)
+
+	callRescan := func(wg *sync.WaitGroup) {
+		defer wg.Done()
+
+		err := nc.Rescan(&startHash, addrs, nil)
+		require.NoError(t, err)
+	}
+
+	// callNotifyReceived calls the NotifyReceived() method and asserts it
+	// completes with no errors.
+	callNotifyReceived := func(wg *sync.WaitGroup) {
+		defer wg.Done()
+
+		err := nc.NotifyReceived(addrs)
+		require.NoError(t, err)
+	}
+
+	// callNotifyBlocks calls the NotifyBlocks() method and asserts it
+	// completes with no errors.
+	callNotifyBlocks := func(wg *sync.WaitGroup) {
+		defer wg.Done()
+
+		err := nc.NotifyBlocks()
+		require.NoError(t, err)
+	}
+
+	// executeCalls launches the wanted number of goroutines, waits
+	// for them to finish and signals all done by closing the returned
+	// channel.
+	executeCalls := func(n int) <-chan struct{} {
+		var (
+			done = make(chan struct{})
+			wg   sync.WaitGroup
+		)
+
+		go func() {
+			defer close(done)
+			defer wg.Wait()
```

Comment:
I think the `wg.Wait()` should come before the `close(done)`, otherwise the `select` below will return even though the wait might have still blocked. Maybe the wait should be right after the loop instead?

Reply:
defer statements are executed from the bottom up, we wait and then close done.